Allow rm_stm to control max compaction offset

ajfabbri opened the pull request:
Allow `rm_stm` to limit maximum offset that compaction may affect, so it can prevent unresolved transactional messages from affecting other records.

Fixes #4823 
## Highlights:
                                                                                                                      
**storage:` add max_collectible_offset to compaction_config**
                                                                                                                      
- Use model::offset::max() as a default value; this means you may compact up to any offset.                                                                                           
- Compaction uses max_collectible_offset, which is set in latter patch in this PR.                                                                                 
                                                                                                                      
### storage/tests: add key-limited random batch generator                                                             
                                                                                                                      
For testing compaction, we want a variant of append_random_batches() which limits the total number of distinct keys used. Unlike fully-random keys, this ensures that compaction can actually do some work.                                 
                                                                                                                      
### s/test/util: basic log gap analysis for compaction tests                                                          
                                                                                                                      
Introduce a utility for compaction testing which helps detect which offset ranges were affected by compaction. This utility makes a single pass over all of the message offsets in a given log instance, and returning the total number of offset gaps, and the min and max offsets that were missing.                                                            
                                                                                                                      
### storage/test: integration test for max compaction (max_collectible) offset                                                          
                                                                                                                      
Three integration tests that verify desired behavior:                                               
                                                                                                                      
- Randomized batch / messsage generaton with a limited number of keys. Verify that compaction only deletes messages less than this max.                                                                                                     
                                                                                                                      
- Single key with max compact offset (MCO): fixed message test that checks that the boundaries of the compacted area are precisely as expected.  
                                                                                                                      
- Single key w/o MCO: same as above, but with max_collectible_offset unset, we validate that compactions works as before.  
                                                                                                                      
### rm_stm: use last stable offset to limit compaction                                                                
                                                                                                                      
Last Stable Offset (LSO) is the largest offset such that no messages, up to and including LSO, are part of an unresolved transaction.                                                                                                                                                                                                                             Use LSO to determine the maximum offset that may be affected by compaction. This ensures that messages for open transactions--which may be aborted or committed in the future--cannot affect compaction. For example, an aborted message cannot replace a previous message with the same key, during compaction.                                                                                                           

### v1 PR
included these changes to support separate control for deletion/eviction versus compaction, which we agreed can be eliminated for now:
~~storage: stm_manager support for compactable_stm~~
 ~~Make stm_manager aware of a new type of stm: compactable_stm. This allows STM implementations to control compaction policy without leaking details into the main compaction logic.~~                                                          
                                                                                                                      
~~For now, this just exposes max_compactable_offset(), which allows rm_stm to prevent unresolved transactional messages from affecting compaction. We also plan to use this plumbing for a per-message predicate which will allow compaction to query the STM, and thus handle things like aborted records properly.~~ 

## Release notes

* none

ajfabbri commented on src/v/storage/segment.h:
```diff
@@ -49,8 +49,12 @@ class segment {
         model::offset committed_offset;
+        // Offset of last message written to this log
         model::offset dirty_offset;
+        // Offset of last message flushed to disk in this log
```
@dotnwat are these descriptions correct?

mmaslankaprv replied:
they are not, log knows nothing about the replication. `committed_offset` is the one that was fsynced, `dirty_offset` is the one that was written to log (may not yet be on disk), stable offset is the one that was written do disk

ajfabbri replied:
Thanks. I thought these should have comments. I will update them.

ajfabbri reviewed (commented):
Needs an integration test; to validate the compaction offset limiting works as expected.

ajfabbri commented on src/v/storage/types.h:
```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
+            vassert(!stm, "Cannot assign two compactable stms");
+            _comp_stm = stm;
+        }
+    }
```
I did some C++. 🎉 😃   Is this the best way to do this?

ajfabbri commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -234,36 +234,34 @@ ss::future<ss::stop_iteration> copy_data_segment_reducer::do_compaction(
-        return ss::make_ready_future<stop_t>(stop_t::no);
+        co_return stop_t::no;
     }
-    return compress_batch(original, std::move(to_copy.value()))
```
🤦 These two diffs are much easier to view on a whitespace-aware viewer, and it makes the reason I split the change into two (for review-ability) clearer.

dotnwat commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -229,37 +229,34 @@ copy_data_segment_reducer::filter(model::record_batch&& batch) {
-                start_offset + header_size);
-          });
-    });
+    auto batch = co_await compress_batch(original, std::move(to_copy.value()));
```
unless you want to suppress clang-format in the previous commit so that the indentation makes the changes really clear, I'd squash this commit into the previous one. I'd probably vote to squash the changes are pretty straight forward.

dotnwat commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -265,12 +265,11 @@ ss::future<ss::stop_iteration>
 copy_data_segment_reducer::operator()(model::record_batch&& b) {
```
I'd recommend taking this opportunity to change the method to take `b` by value since this is a coroutine now. code is ok, but taking references in coroutines makes them prone to dangling reference bugs (same with the previous commit's changes).

ajfabbri replied:
Added another commit towards the end for this, thank you.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -122,6 +122,7 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
         co_await current_log.handle.compact(compaction_config(
           collection_threshold,
           _config.retention_bytes(),
+          current_log.handle.stm_manager()->max_compactible_offset(),
```
it kind of seems like rm_stm should update storage layer (as opposed to storage querying rm_stm). rm_stm is in a position to know when an offset changes, and having information flow downward there is less complex coupling. wdyt?

rystsov replied:
@mmaslankaprv suggested using polling, it seems like a more straightforward solution, also in the next part of the compaction /transaction integration the compaction logic will need to query if a given record is a part of the aborted or committed transaction. using an object to query instead of passing a list of aborted records gives more space for memory optimization

dotnwat replied:
I wonder if we are starting to approach a situation where we need to think about moving part of the transaction management into the storage layer.

dotnwat replied:
> @mmaslankaprv suggested using polling, it seems like a more straightforward solution

I'm not sure I agree it is more straight forward when we consider the lifetime of all the new objects involved. But I do agree for querying and memory optimization of aborted records it doesn't make sense to push that down. However, I do think that suggests some reorganization would be beneficial.

rystsov replied:
I agree that the inverted dependencies are annoying and it's a good idea to lift compaction logic above raft and let stm directly control it but it's out of scope of the compaction / transactions integration and we can't condition transactions release on the fundamental refactoring of the compaction logic.

Right now we compact adjacent segments but when the transactions are involved in order to decide if a record b overwrites a record a (even when the records are from the same segment) we may need to look ahead for a unbounded amount of segments for search for corresponding abort or commit txn markers. So we either need to change compaction from working with adjacent segments to work with unbounded segments or to rely on an external oracle who happens to know future to guide the compaction logic. The oracle requires less changes.

But there is a silver lining - we are not increasing the tech debt balance. The storage layer already depends on stm_manager which contains a reference to rm_stm so things are not getting worse.

mmaslankaprv replied:
I think i agree with Noah here, it would be cleaner to reverse the dependency, but i also got Denis'es point about already existing `stm_manager`

ajfabbri replied:
Yep. I was thinking similar things as I wrote this PR; would like to chat about layering here.

dotnwat replied:
> But there is a silver lining - we are not increasing the tech debt balance. The storage layer already depends on stm_manager which contains a reference to rm_stm so things are not getting worse.

thanks for mentioning this observation

rystsov commented on src/v/storage/types.h:
```diff
@@ -353,6 +356,8 @@ struct compaction_config {
     // remove one segment if log is > max_bytes
     std::optional<size_t> max_bytes;
+    // Cannot compact past this offset (i.e. for unresolved txn records)
+    model::offset max_compactable_offset;
```
please add a comment if we allow to compact strictly less or less-or-equal than `max_compactable_offset`

ajfabbri replied:
Will clarify. For the recod: "max compactable offset" to me means that it is the maximum value for which the condition (compactable) holds. "Cannot compact past this offset" also implies `>`

rystsov reviewed (commented):
Looks good but I suggest to postpone merging before we start using max_compactable_offset to actually limit compaction to get the whole picture of the change.

rystsov commented on src/v/storage/types.h:
```diff
@@ -340,11 +340,14 @@ struct compaction_config {
     explicit compaction_config(
       model::timestamp upper,
       std::optional<size_t> max_bytes_in_log,
+      std::optional<model::offset> max_compact_offset,
```
we already use `model::offset::max()` as an unset value (in stm_manager) so it's better to stick with it and not to introduce `optional`. making a type optional adds another dimension to the value space which we should keep in mind when we use the type.

ajfabbri replied:
This particular API, `compaction_config` uses `optional`, so I thought it was more consistent. I did use `offset::max()` for the underlying value, though. I feel like nullopt for the test cases is more ergonomic, but I can change it to `...max()`..

dotnwat replied:
> we already use model::offset::max()

most of the code base uses `model::offset{}` (aka `::min`) to indicate an unset state. unless using `::max` is convenient because of a calculation that is based on `min(a, b)` then it is probably worth sticking to default initialization?

mmaslankaprv commented on src/v/storage/types.h:
```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
 public:
-    void add_stm(ss::shared_ptr<snapshotable_stm> stm) { _stms.push_back(stm); }
+    template<
+      typename T,
```
no need for this to be generic, just pass in `ss::shared_ptr<compactible_stm>`

mmaslankaprv replied:
`ss::shared_ptr` supports virtural dispatch

dotnwat replied:
Agree. Afaict the stm type hierarchy can already handle this without generics. Or create `add_compactible_stm`?

ajfabbri replied:
I took another stab in force-push. We can continue discussion there.

mmaslankaprv commented on src/v/storage/types.h:
```diff
@@ -118,8 +139,20 @@ class stm_manager {
+    // Since compactable_stm is a snapshotable_stm, it will appear in this list
+    // as well as in the value for _comp_stm.
     std::vector<ss::shared_ptr<snapshotable_stm>> _stms;
+    ss::shared_ptr<compactable_stm> _comp_stm;
```
do we need a special case for this stm type ?, would it be enough to use `max_collectible_offset` as a compaction boundary ?

ajfabbri replied:
This was one of my first ideas. @rystsov indicated that this compaction codepath should be distinct from the snapshot/eviction code. We should chat about the definition of "max collectible" and all the use cases.

ajfabbri replied:
IIUC, he's saying the value of having them separate is that it allows stm to control compaction and eviction separately, and asking if we think it is worthwhile.

mmaslankaprv replied:
i would vote for using `max_collectible_offset` in compaction, there is no point in compacting a log if it is about to be delete either way, having a special case for compaction stm make the whole thing more complicated

ajfabbri replied:
As much as I'd love to get this merged, I tend to agree with @mmaslankaprv. At least this is what I wanted to try first, because we can always add separate compaction control in the future if needed. I didn't see a need for this currently, but maybe in the future we introduce a new compaction algorithm or something that needs it?

Unless there are any objections, I'll whip up a new PR or add a patch which eliminates the separate compactable offset plumbing, so we can see the difference in approaches.

ajfabbri commented on src/v/storage/types.h:
```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
+        using type = std::decay_t<T>;
+        _stms.push_back(stm);
+        if constexpr (std::is_same_v<type, compactable_stm>) {
+            vassert(!stm, "Cannot assign two compactable stms");
```
Good catch. :+1:

ajfabbri replied:
Updated in force-push, resolving

dotnwat commented on src/v/storage/types.h:
```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
-    void add_stm(ss::shared_ptr<snapshotable_stm> stm) { _stms.push_back(stm); }
+    template<
+      typename T,
+      typename = std::enable_if_t<std::is_base_of_v<snapshotable_stm, T>>>
```
I don't think you need SFINAE here

ajfabbri replied:
Yep. Originally thought they were going to be distinct types but they ended up having a common base. Consider this deleted practice code for me.

dotnwat commented on src/v/storage/types.h:
```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
+    template<
+      typename T,
+      typename = std::enable_if_t<std::is_base_of_v<snapshotable_stm, T>>>
+    void add_stm(ss::shared_ptr<T> stm) {
```
It might be a clearer api if you added an `add_compactable_stm` interface since we already have a run-time requirement that two compatible stms cannot be added

ajfabbri replied:
Agreed.

ajfabbri replied:
I ended up not doing this (separately-named method) because of the awkwardness of the class hierarchy; a `compactable_stm` is not a `snapshotable_stm`. Ended up using a dynamic cast. LMK what you think.

dotnwat commented on src/v/storage/types.h:
```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
+    void add_stm(ss::shared_ptr<T> stm) {
+        using type = std::decay_t<T>;
+        _stms.push_back(stm);
+        if constexpr (std::is_same_v<type, compactable_stm>) {
```
is `is_same` what we want here? it looks to me like `type` is going to be part of the type hierarchy not necessary exactly compatble_stm?

ajfabbri replied:
Yeah, I removed this code, but I noticed it should have been, what, `is_base_of<>`?

dotnwat replied:
yes, probably. still, it's pretty rare to need such a thing, and is a red flag here that the interface / types can be improved.

ajfabbri commented:
Force push:
- Remove template for add_stm(). Could also add separate method `add_compactable_stm()`, but that either requires a dynamic cast (`compactable_stm` is not a `snapshotable_stm`), or calling both `add_stm()` and `add_compactable_stm()` in the case of an instance of `rm_stm`.
- Use offset::max() instead of optional/nullopt
- Clarify `>=` for max collectable offset

ajfabbri commented:
Marking draft--will also include integration with compaction and a test in this PR. Stay tuned.

ajfabbri commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -2110,3 +2110,71 @@ FIXTURE_TEST(test_querying_term_last_offset, storage_test_fixture) {
+      new_batches.begin(),
+      new_batches.end(),
+      size_t(0),
+      [](size_t acc, model::record_batch& b) { return acc + b.size_bytes(); });
```
This `total_size` calculation will be removed (assuming I don't use it).

ajfabbri commented:
Force push: Finish getting integration test working, fixing random batch (with limited key set) generation, removing debugging code. Next force-push will be a rebase on latest dev (sorry) since there is a conflict.

ajfabbri commented:
Force-push: resolve upstream merge conflict in rm_stm.h.

ajfabbri commented:
CI failures are unrelated: #5373, #5417, #5463

ajfabbri commented:
This PR is still ready for review. PR failures marked above (a lot at the moment--since we're bringing in good stuff like upgrade testing). Understand folks are busy at the moment--no worries, been there. :smiley:

rystsov commented on src/v/storage/tests/utils/log_gap_analysis.h:
```diff
@@ -0,0 +1,36 @@
+// Utility to analyze offset gaps in logs, for testing compaction logic.
+
+struct log_gap_analysis {
+    model::offset first_gap_offset{-1};
```
Can you please add a comment when `first_gap_offset` and `last_gap_offset` mean?

ajfabbri replied:
Added a comment explaining the fields, thanks. Update coming soon.

rystsov commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -2129,3 +2132,87 @@ FIXTURE_TEST(test_querying_term_last_offset, storage_test_fixture) {
+    auto log = mgr.manage(std::move(ntp_cfg)).get0();
+    auto disk_log = get_disk_log(log);
+
+    // (1) append some random data, with limited number of distinct keys, so
```
Can we have at least a single test with a predefined fixed set of keys? With randomness it's impossible to reproduce a failing test, also it may happen that all the keys are different and compaction happened beyond the allowed boundary but because all the keys are distinct it had zero effect and the test wouldn't fail when it should.

Before we use randomness we should be certain that the basic scenarios work:

- defined max_compactable_offset blocks compaction
- default max_compactable_offset allows compaction to happen

ajfabbri replied:
For sure, happy to add a simple, deterministic case as well.

ajfabbri replied:
I've added two additional test cases without the message randomization, with and without setting the max compactable offset.

graphcareful commented on src/v/random/random_test.cc:
```diff
@@ -34,3 +35,19 @@ BOOST_AUTO_TEST_CASE(alphanum_generator) {
+    for (auto i = 0; i < 100; i++) {
+        auto s = random_generators::gen_alphanum_max_distinct(cardinality);
+        // ensure no \0 in size
+        for (auto j = 0; j < random_generators::alphanum_max_distinct_strlen;
```
does `alphanum_max_distinct_strlen` need to be public just so it can be used here? Can you just iterate through the string checking for \0 ?

ajfabbri replied:
Sure, we just lose testing the property that there aren't any unexpected null characters, and the ability of callers to know the maximum length.

ajfabbri commented on src/v/cluster/persisted_stm.h:
```diff
@@ -67,7 +67,7 @@ struct stm_snapshot {
  *
  * When persisted_stm becomes a leader it syncs its state with the tip
- * of the log (by writing a checkout batch checkpoint_batch_type and
+ * of the log (by writing a checkpoint batch and
```
Thanks. I took another stab at updating this comment in the next revision (force push coming soon).

ajfabbri commented on src/v/model/tests/random_batch.cc:
```diff
@@ -46,10 +46,14 @@ std::vector<model::record_header> make_headers(int n = 2) {
 
-model::record
-make_random_record(int index, std::optional<size_t> rec_size = std::nullopt) {
+model::record make_random_record(
```
There is only one caller currently--but I went ahead and made this change. Will also make these helper functions static. Updated PR coming soon.

ajfabbri commented on src/v/random/generators.cc:
```diff
@@ -32,4 +32,18 @@ ss::sstring gen_alphanum_string(size_t n) {
     return s;
 }
 
+ss::sstring gen_alphanum_max_distinct(size_t cardinality) {
```
The idea is often called a `generator`; a function that generates a set, one entry at a time.

ajfabbri commented on src/v/random/generators.cc:
```diff
@@ -0,0 +1,35 @@
+/*
```
Agreed. I think I met that criteria, except for the part where I pulled `std::string_view chars` up to file scope. Good catch. :eyes:

dotnwat commented on src/v/random/generators.cc:
```diff
@@ -32,4 +32,18 @@ ss::sstring gen_alphanum_string(size_t n) {
+    auto s = ss::uninitialized_string(alphanum_max_distinct_strlen);
+    std::generate_n(s.begin(), alphanum_max_distinct_strlen, [&] {
+        auto c = chars[next_index];
+        next_index = (next_index + key_num) % num_chars;
```
seems to me this is high _non_-random. it seems like it might be better to choose random() % cardinality so that you are selecting randomly within a range whose size is constrained by cardinatility.

ajfabbri replied:
My goal is to choose k arbitrary keys.

ajfabbri replied:
Also feel free to share a code snippet. Not sure the details of what you are proposing other than keys in [0,n) with the same distribution but integral values.

dotnwat replied:
what i'm saying is that in your test if you want key cardinality 100 then you could use `fmt::format("{}", get_int(100));`

ajfabbri replied:
It's a good point. I can do a larger fmt template for bigger keys. I was originally going to make max len a variable too, but just wanted a free function that acted like a generator, and having to use the same max size and max cardinality args each time to get the same set felt awkward.

dotnwat commented on src/v/random/generators.h:
```diff
@@ -35,6 +35,9 @@ static thread_local std::default_random_engine gen(internal::get_seed());
 ss::sstring gen_alphanum_string(size_t n);
 
+static constexpr size_t alphanum_max_distinct_strlen = 32;
+ss::sstring gen_alphanum_max_distinct(size_t max_cardinality);
```
please write a comment here about what this is doing.

ajfabbri replied:
Covered in another comment and addressed. PR update is in progress--expect update soon

dotnwat commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -234,7 +234,7 @@ copy_data_segment_reducer::filter(model::record_batch&& batch) {
 
 ss::future<ss::stop_iteration> copy_data_segment_reducer::do_compaction(
-  model::compression original, model::record_batch&& b) {
+  model::compression original, model::record_batch b) {
```
this can be squashed into previous commits that switch methods to coroutines.

ajfabbri replied:
I have squashed.

dotnwat commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -232,41 +232,36 @@ copy_data_segment_reducer::filter(model::record_batch&& batch) {
 }
+
 ss::future<ss::stop_iteration> copy_data_segment_reducer::do_compaction(
   model::compression original, model::record_batch&& b) {
```
EDIT: i see it's now at the end of the PR. should be easy to sqash that into the first and second commits?

ajfabbri replied:
Yep. Squashed it. Thanks again.

ajfabbri replied:
This is done, resolving.

ajfabbri commented on src/v/cluster/rm_stm.h:
```diff
@@ -172,6 +175,10 @@ class rm_stm final : public persisted_stm {
       aborted_transactions(model::offset, model::offset);
 
+    inline model::offset max_compactable_offset() override {
+        return last_stable_offset();
```
Agreed. I'm splitting this out to a separate commit in the next update.

ajfabbri replied:
Now in separate commit with explanatory message.

ajfabbri commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -349,6 +349,19 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
 
+        // since we don't support partially-compacted segments, a segment must
+        // end before the max compactable offset to be eligible for compaction.
+        if (
+          seg->offsets().stable_offset
+          > _stm_manager->max_compactible_offset()) {
+            vlog(
+              gclog.debug,
+              "[{}] segment {} is past compactible offset, skipping.",
+              config().ntp(),
+              seg->reader().filename());
+            continue;
+        }
+
```
Will do, thanks.

ajfabbri commented:
**(Pending--will continue tomorrow)** ~~Force push~~--addressing review comments:
- Rebase on dev to fix conflicts.
- Add a second, non-randomized, compaction integration test (good for catching off-by-one's and such).
- Add comments to gap analysis (helper for compaction tests). Change result struct slightly; track min and max missing offsets (i.e. min and max offsets that were deleted by compaction).
- Squash "pass by value" patch with co-routine-ization patches. Squash compaction config commit with change to compaction code.
- Separate out final "integration" patch as originally planned; the decision to use last_stable_offset() is its own commit.
- Add more detail to commit messages, where needed.
- Add an overload for record generator function, instead of multiple optionals.

ajfabbri commented:
Current issue with my strict integration tests is that the first segment is not always compacted. I'm looking into it.

ajfabbri commented:
FWIW, keys generated by `gen_alphanum_max_distinct()` look like this (pasted below). Keep in mind this is only a couple of lines of code and just for testing--so it doesn't matter a ton.
```
wSfBXkG2pL7uQdzViE0nJ5sObxTgCYlH                                                                                                                                                                                                              
UFqbVGrcWHsdXIteYJufZKvg0Lwh1Mxi                                                                                                                                                                                                              
yWlJ7wUjH5uShF3sQfD1qOdBZoMbzXmK                                                                                                                                                                                                              
51XTPLHDzvrnjfb62YUQMIEAwsokgc73                                                                                                                                                                                                              
630XUROLIFCzwtqnkheb741YVSPMJGDA                                                                                                                                                                                                              
pET8oDS7nCR6mBQ5lAP4kzO3jyN2ixM1                                                                                                                                                                                                              
tM5pI1lEXhATdwP8sL4oH0kDWgzScvO7                                                                                                                                                                                                              
myKW8lxJV7kwIU6jvHT5iuGS4htFR3gs                                                                                                                                                                                                              
HfMkRpWu1z6EcJhOmTrYw3B8GeLjQoVt                                                                                                                                                                                                              
zYoNdC1rQgF4uTjI7xWmLbAZpOeD2sRh                                                                                                                                                                                                              
D6A3x0uXrUoRlOiLfIcF8C5z2wZtWqTn                                                                                                                                                                                                              
2VOHAtmf70TMFyrkd5YRKDwpib3WPIBu                                                                                                                                                                                                              
7531ZXVTRPNLJHFDBzxvtrpnljhfdb86                                                                                                                                                                                                              
IhPoWv3CbJiQpXw4DcKjRqYx5EdLkSrZ                                                                                                                                                                                                              
YNCrg4TIxmbZODsh5UJync0PEti6VKzo                                                                                                                                                                                                              
Nr4ImZDhUycPt6Ko1FjWAeRv8Mq3HlYC                                                                                                                                                                                                              
qGWdtJZgwM2jzP5mCS8pFVcsIYfvL1iy                                                                                                                                                                                                              
GdJgMjPmSpVsYv1y4B7EbHeKhNkQnTqW                                                                                                                                                                                                              
0RIzqh7YPGxof5WNEvmd3ULCtkb1SJAr                                                                                                                                                                                                              
rIZhyP6oFWevM3lCTbsJ0izQ7pGXfwN4                                                                                                                                                                                                              
oCQ4jxLZesGU8nBP3iwKYdrFT7mAO2hv                                                                                                                                                                                                              
vQcxSezUgBWiDYkF0mH2oJ4qL6sN8uPb                                                                                                                                                                                                              
3XRLFztnhb4YSMGAuoic5ZTNHBvpjd60                                                                                                                                                                                                              
lwHS3fqBMX8kvGR2epALW7juFQ1dozKV                                                                                                                                                                                                              
WJwj5SFsf1OBobXKxk6TGtg2PCpcYLyl                                                                                                                                                                                                              
KlVw6HiSt3EfPq0BcMnXy8JkUv5GhRs2                                                                                                                                                                                                              
4ZUPKFAvqlgb50VQLGBwrmhc61WRMHCx                                                                                                                                                                                                              
fkpuzEJOTY38ejotyDINSX27dinsxCHM                                                                                                                                                                                                              
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa                                                                                                                                                                                                              
A0rRiI8zZqQhH7yYpPgG6xXoOfF5wWnN                                                                                                                                                                                                              
RzhYGo5NvdUCk1Jr8QygXFn4MucTBj0I                                                                                                                                                                                                              
Mp1EhTw8Lo0DgSv7KnZCfRu6JmYBeQt5                                                                                                                                                                                                              
sK2lDVewO6pHZiASbtL3mEWfxP7qI0jB                                                                                                                                                                                                              
cegikmoqsuwyACEGIKMOQSUWY02468bd                                                                                                                                                                                                              
nAN0erER4ivIV8mzMZdqDQ3huHU7lyLY                                                                                                                                                                                                              
Ot7Mr5Kp3In1GlZEjXChVAfTydRwbPu8                                                                                                                                                                                                              
iqyGOW4dltBJRZ7gowEMU2bjrzHPX5em                                                                                                                                                                                                              
E8D7C6B5A4z3y2x1w0vZuYtXsWrVqUpT                                                                                                                                                                                                              
XLznbYMAocZNBpd0OCqe1PDrf2QEsg3R                                                                                                                                                                                                              
uO8tN7sM6rL5qK4pJ3oI2nH1mG0lFZkE                                                                                                                                                                                                              
kuEOY8jtDNX7isCMW6hrBLV5gqAKU4fp                                                                                                                                                                                                              
JjSs1BbKkTt2CcLlUu3DdMmVv4EeNnWw                                                                                                                                                                                                              
jsBKT2cluDMV4enwFOX6gpyHQZ8irAJS                                                                                                                                                                                                              
1TLDvnf6YQIAskc3VNFxph80SKCume5X                                                                                                                                                                                                              
gmsyEKQW28flrxDJPV17ekqwCIOU06dj                                                                                                                                                                                                              
xUiF2qNbyVjG3rOczWkH4sPdAXlI5tQe                                                                                                                                                                                                              
C4xZsUnPiKdF7A2vXqSlNgIbD5y0tVoQ                                                                                                                                                                                                              
B2uVnOgH8A1tUmNfG7z0sTlMeF6yZrSk                                                                                                                                                                                                              
eimquyCGKOSW048dhlptxBFJNRVZ37cg                                                                                                                                                                                                              
LnYAcNp0CePr2EgRt4GiTv6IkVx8KmXz                                                                                                                                                                                                              
TDn6QAk3Nxh0KueXHrbUEo7RBl4Oyi1L                                                                                                                                                                                                              
dgjmpsvyBEHKNQTWZ258cfiloruxADGJ                                                                                                                                                                                                              
SBk2LudVEn5OxgYHq8RAj1KtcUDm4Nwf                                                                                                                                                                                                              
VHtf0Myk5RDpbWIug1Nzl6SEqcXJvh2O                                                                                                                                                                                                              
bcdefghijklmnopqrstuvwxyzABCDEFG                                                                                                                                                                                                              
PvbQwcRxdSyeTzfUAgVBhWCiXDjYEkZF                                                                                                                                                                                                              
FbGcHdIeJfKgLhMiNjOkPlQmRnSoTpUq                                                                                                                                                                                                              
ZPFvlb0QGwmc1RHxnd2SIyoe3TJzpf4U                                                                                                                                                                                                              
QxeUBiYFm2Jq6NubRyfVCjZGn3Kr7Ovc                                                                                                                                                                                                              
876543210ZYXWVUTSRQPONMLKJIHGFED                                                                                                                                                                                                              
hovCJQX4cjqxELSZ6elszGNU18gnuBIP
```

dotnwat commented:
restarted ci

ajfabbri commented:
Force pushes: Rebase on latest dev to fix significant conflicts with #5991. Also fixed comments per Michal's suggestion.

ajfabbri commented:
Force-push: 
- Eliminate `max_compactible_offset`. Use existing `max_collectible_offset` to limit both deletion and compaction.
- Eliminate `compactible_stm` and associated changes.
- Rebase on latest dev.

ajfabbri commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -349,6 +349,28 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+              segment->reader().filename(),
+              segment->offsets().stable_offset,
+              cfg.max_collectible_offset);
+            continue;
```
This needs to be fixed. The surrounding code was rewritten and this first attempt at merging is incorrect--the `continue` here will be an infinite loop. Will update shortly.

ajfabbri replied:
Done.

ajfabbri commented:
Setting to draft while I fix merge w/ recent do_compaction() changes.. Stay tuned.

ajfabbri commented:
Force-push: 
- Fix compaction config patch's do_compact() logic to work with latest code.
- Fix final assert in max_compact_offset tests to not be so strict. Add comment explaining the situation with adjacent segment compaction being LIFO instead of FIFO

mmaslankaprv commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -333,15 +333,39 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+    auto offsets_compactible = [&](segment& s) {
+        if (log::has_compactible_offsets(s, cfg)) {
+            vlog(
+              gclog.debug,
```
trace maybe, there may be hundreds of thousands of segments

ajfabbri replied:
Good idea. (I think it didn't print that much spam in my 2000 partition test because it exits after it processes one segment.)

ajfabbri replied:
I kept this one as `debug`, because one debug-level message each time we compact a segment is quite reasonable. The other case is trace now. .

ajfabbri commented:
Force-push:
- Rebase on dev.
- Move "skipping non-compactible segment" logging to TRACE, but leave per-compaction one as DEBUG (since our compaction rate is limited, so are these log messages).
- Fix backwards conditional I added during rebase

mmaslankaprv commented on src/v/storage/log.h:
```diff
@@ -181,13 +182,24 @@ class log final {
         return _impl->set_collectible_offset(o);
     };
 
+    /**
+     * We currently use `max_collectible_offset` to control both
+     * deletion/eviction, and compaction.
+     */
+    static bool
+    has_compactible_offsets(const segment& seg, const compaction_config& cfg) {
+        // since we don't support partially-compacted segments, a segment must
+        // end before the max compactible offset to be eligible for compaction.
+        return seg.offsets().stable_offset <= cfg.max_collectible_offset;
+    }
+
```
Maybe extract to util function this is not really related to log, we should not use segments in the `log` abstraction as segments are valid only for disk_log_impl

mmaslankaprv replied:
maybe make it a `segment` method ?

ajfabbri replied:
Ah.. I see. I thought segment was generic as well. I like moving it to `segment`, thanks.

ajfabbri replied:
Done in latest update.

mmaslankaprv commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -334,15 +334,39 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+        if (log::has_compactible_offsets(s, cfg)) {
+            vlog(
+              gclog.debug,
+              "[{}] segment {} stable offs {}, max compactible {} -> compact.",
```
nit: would be great if you could change this message to be more descriptive. now the `->` doesn't really mean anything without having a context Would be better if it says:
```
compacting segment
```

ajfabbri replied:
`->` is mathematical notation for "implies". I can change it to "compacting" but "segment" is redundant IMHO

ajfabbri replied:
Changed to "compacting" in latest update.. thanks for the suggestion.

mmaslankaprv commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -334,15 +334,39 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
-    // find first not compacted segment
+
+    // create a logging predicate for offsets..
+    auto offsets_compactible = [&](segment& s) {
```
do we only need `cfg` reference here ?

ajfabbri replied:
Yep. Do you prefer explicit `&cfg`? I couldn't think of an example of where default capture by reference *from* a coroutine is bad. Obviously capturing a reference to a _non-coroutine_ automatic var is bad, but this is a coroutine. Maybe good hygiene to always be explicit? The lambda its self is not a coroutine.

dotnwat replied:
the code is safe. but generally the code base (the exception being tests) avoids capturing the entire context for hygiene reasons.

ajfabbri replied:
I'm cool with either. I love concise things but yeah, in non-test code extra hygiene is good. I'll update this, thank you.

BenPope replied:
I also like being explicit for captures, lifetimes and the cost of copies should be explicit.

mmaslankaprv commented on src/v/cluster/rm_stm.cc:
```diff
@@ -247,6 +247,7 @@ rm_stm::rm_stm(
 }
 
 bool rm_stm::check_tx_permitted() {
+    // TODO support compaction
```
unrelated change

ajfabbri replied:
seems related to me. Did you want this pulled out to a separate commit?

ajfabbri replied:
I'll pull the comments out into a separate patch now, resolving.

mmaslankaprv commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -2780,11 +2780,18 @@ do_compact_test(const compact_test_args args, storage_test_fixture& f) {
+
+    // If adjacent segment compaction worked in order from oldest to newest, we
+    // could use this assert.
+#if 0
```
can we simply remove this if the assertion is incorrect ? also it should be squashed with previous changes to keep history bisectable

ajfabbri replied:
Did you see the commit comment for this patch? I think this is useful documentation.. It took a while to figure out the behavior here. I agree on the squishing part--I pulled this out for reviewers' benefit.

ajfabbri replied:
LMK if you still want the comment removed.

dotnwat replied:
this seems reasonable, but using a multi-line comment instead of an if 0 block would be more idiomatic in the code base. this particular case seems very clear and the if 0/endif kind of highlights the aspect of it being code that the comment is referring to

rystsov replied:
`#if false` feels like a merge-broken version check or something, agree with Noah, it's better to use a multi-line comment

ajfabbri commented on src/v/storage/log.h:
```diff
@@ -17,6 +17,7 @@
 #include "seastarx.h"
 #include "storage/log_appender.h"
 #include "storage/ntp_config.h"
+#include "storage/segment.h"
```
Log and Segment are both generic.. that is why I put it here. Let me take another look.

ajfabbri replied:
We should rename `log::segment_count()` to `log::mystery_number()`.. then `log` will _really_ know nothing about underlying segments. :laughing:

ajfabbri commented:
Force-push:
- Move `has_compactible_offsets()` to be a method on segment.
- Clarify log message "compacting"

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -334,15 +334,39 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
-    // find first not compacted segment
+
+    // create a logging predicate for offsets..
+    auto offsets_compactible = [&](segment& s) {
```
the code is safe. but generally the code base (the exception being tests) avoids capturing the entire context for hygiene reasons.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -334,15 +334,39 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
         const auto end_it = std::prev(_segs.end());
         auto seg_it = std::find_if(
-          _segs.begin(), end_it, [](ss::lw_shared_ptr<segment>& s) {
+          _segs.begin(), end_it, [&](ss::lw_shared_ptr<segment>& s) {
```
you can capture the lambda in this lambda capture.

ajfabbri replied:
Thanks. Might be good to add a linter rule for default reference captures?

dotnwat commented on src/v/storage/segment.h:
```diff
@@ -295,6 +298,14 @@ inline size_t segment::size_bytes() const {
+segment::has_compactible_offsets(const compaction_config& cfg) const {
+    // since we don't support partially-compacted segments, a segment must
+    // end before the max compactible offset to be eligible for compaction.
+    return offsets().stable_offset <= cfg.max_collectible_offset;
```
would it possible that we want dirty offset since it's generally going to be the highest possible offset if writes happen to be in flight?

ajfabbri replied:
I thought we only compact closed segments?

dotnwat commented on src/v/storage/tests/storage_e2e_test.cc:
```diff
@@ -1955,7 +1994,11 @@ FIXTURE_TEST(changing_cleanup_policy_back_and_forth, storage_test_fixture) {
     BOOST_REQUIRE_EQUAL(disk_log->segment_count(), 3);
 
     storage::compaction_config c_cfg(
-      model::timestamp::min(), std::nullopt, ss::default_priority_class(), as);
+      model::timestamp::min(),
+      std::nullopt,
+      model::offset::max(),
+      ss::default_priority_class(),
+      as);
```
it looks like further down in the file there are other compaction configs that need updated too?

ajfabbri replied:
I believe I misplaced one chunk during one of the rebases. I'm making sure this config change can be bisected now, thanks.

dotnwat commented on src/v/model/tests/random_batch.h:
```diff
@@ -36,6 +36,9 @@ struct record_batch_spec {
 };
 
+// Makes an random alphanumeric string, encoded in an iobuf.
+iobuf make_iobuf(size_t);
```
can we change the name of this. it looks like it had been intended to be internal to the translation unit, but now at the global scope the name isn't helpful.

ajfabbri replied:
Isn't the effective name here `random_generators:make_iobuf`?

dotnwat replied:
it looks to me like the fully qualified name is `model::test::make_iobuf`. i'd move it to random generators or bytes test helpers.

ajfabbri replied:
Ah, I was confusing `using-directive` with `using-declaration` behavior here. Will do, thanks.

dotnwat commented on src/v/cluster/rm_stm.h:
```diff
@@ -174,6 +174,10 @@ class rm_stm final : public persisted_stm {
     ss::future<std::vector<rm_stm::tx_range>>
       aborted_transactions(model::offset, model::offset);
 
+    inline model::offset max_collectible_offset() override {
```
i think that the inline keyword here is redundant/implied

ajfabbri replied:
Oops. copy-pasta. :ramen: Thanks!

dotnwat reviewed (commented):
nice stuff. this generally looks good to me.

1. i think one of the earlier commits might not compile
2. added a few minor nits i saw along the way

I was surprised to see the dynamic_cast / interface mixin disappear. I looked back through some of the feedback but didn't see that discussion. Do you have a pointer or could explain?

ajfabbri commented on src/v/random/random_test.cc:
```diff
@@ -34,3 +35,19 @@ BOOST_AUTO_TEST_CASE(alphanum_generator) {
+        }
+        strings.emplace(s);
+    }
+    auto distinct = strings.size();
```
unused variable here. Fixing.

ajfabbri commented:
Force-push:
- rebase on latest dev
-  Fix a warning.

Some of the compaction code has been changing under this PR. Looking at CI failures--some strict compaction  tests I added that were passing, no longer are. Will trace this down tomorrow.

ajfabbri commented:
Force-push: Rebase on dev and fix a default initialization offset in the test that should have been max().

ajfabbri commented:
Force-push:
- Remove redundant `inline`
- Remove two reference default lambda captures, make explicit.
- Rebase on dev.
- Fix formatting.

(Sorry for the multiple pushes--I will debounce better in the future :sweat_smile:)

ajfabbri commented:
> nice stuff. this generally looks good to me.
> 
> 1. i think one of the earlier commits might not compile
> 2. added a few minor nits i saw along the way
> 
> I was surprised to see the dynamic_cast / interface mixin disappear. I looked back through some of the feedback but didn't see that discussion. Do you have a pointer or could explain?

Thanks @dotnwat. The bottom of the PR description covers the change.

ajfabbri commented:
Force-push: should show as empty since end result is the same:
- Split compaction config patch into multiple: 1. add max offset field to config, 2. change compaction to honor the config, 3. Remove unused include. 4. Comments
- Fix bisect-ability for #1 above (chunk was misplaced in later commit when rebasing).

ajfabbri commented:
Debug build CI failure is unrelated #6237..

dotnwat commented:
> random_generators::make_random_iobuf()

good point, i didn't notice it was in this namespace!

dotnwat commented on src/v/storage/tests/utils/disk_log_builder.h:
```diff
@@ -26,7 +26,6 @@
 #include <seastar/core/sstring.hh>
 #include <seastar/util/bool_class.hh>
 
-#include <optional>
```
later in the file:

```c++
      std::optional<size_t> max_partition_retention_size);
```

dotnwat reviewed (approved):
lgtm.

we should have @rystsov (the author of the feature request), and/or @mmaslankaprv approve too.

LenaAn commented on src/v/storage/tests/utils/log_gap_analysis.h:
```diff
@@ -0,0 +1,49 @@
+ * minimum and maximum offsets that were missing.
+ * This is useful for observing which offsets were affected by compaction.
+ */
+struct log_gap_analysis {
```
nit: another useful thing would be the analysis of records read and unique keys read. For example, we produce 10000 records with 200 unique keys and consume 200 records with 200 unique keys. So we see how much we compact data.

LenaAn reviewed (approved):
great change! just small nits

rystsov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -251,6 +251,7 @@ rm_stm::rm_stm(
 }
 
 bool rm_stm::check_tx_permitted() {
+    // TODO support compaction
```
It already exists https://github.com/redpanda-data/redpanda/issues/4824

dotnwat replied:
>can you please open a github issue with a context what exactly need to be done? and put the link here

thanks @LenaAn it is nice to have the link to the github issue in the code comment itself. it's usually a big lift to reverse index from a code comment back to a github review comment.

rystsov commented on src/v/random/generators.h:
```diff
@@ -33,8 +33,24 @@ static thread_local std::default_random_engine gen(internal::get_seed());
+ */
 ss::sstring gen_alphanum_string(size_t n);
 
+static constexpr size_t alphanum_max_distinct_strlen = 32;
```
nit: as i understand `gen_alphanum_max_distinct` returns random string with fixed length equal to  `alphanum_max_distinct_strlen`; because the string length is fixed it's better to rename it to `alphanum_distinct_strlen`

rystsov reviewed (commented):
Looks good! I really like the cover letter, the commit messages and the comments!

ajfabbri commented:
Thanks for the reviews! I will address remaining nits (change preprocessor conditional to comment, etc) in followup PR coming shortly.
